### PR TITLE
Enable switchlayer change in subfolder

### LINKED ISSUE #265: switchlayer not changing when in a subfolder
The following layer search code will need to be made to be recursive to check for overlays within overlays: https://github.com/compassinformatics/cpsi-mapview/blob/master/app/factory/Layer.js#L926

### PR DESCRIPTION
**Pull request by JakobMiksch**
Resolves #265 

Warning: the diff from GitHub looks horrible. 
Actually, only the for-loop has been converted to a recursive function. Hence, the real change is not that big.

I noticed the layertree does not properly update the switch layer anymore. This seemed to be introduced with https://github.com/compassinformatics/cpsi-mapview/commit/081b0e71ca1d1e729bfedaf2cc964f6723291f1e

### CODE REVIEW

**Comment by chrismayer** [comment]
> I noticed the layertree does not properly update the switch layer anymore. This seemed to be introduced with 081b0e7

Can confirm this. Seems like an unrecognized side effect of #289. I'll have a deeper look at this but surely unrelated to this PR.

**Comment by chrismayer** [comment]
@JakobMiksch The tree node should update properly with the fix provided in #322 (please test). Could you add this fix to your PR? Or how should we handle that?

**Comment by JakobMiksch** [comment]
@chrismayer yes, I am trying to include your changes into this pull request as well. Has not worked yet, but I am tracing the problem ...

**Comment by chrismayer** [comment]
@JakobMiksch I created a PR against your branch for this: https://github.com/JakobMiksch/cpsi-mapview/pull/1 Please review and check if this does the trick for you.

**Comment by JakobMiksch** [comment]
@chrismayer unfortunately it does not work yet. The fix works on the top level, but in subgroups it only changes the label of the layertree once, and then not anymore at all.

**Comment by JakobMiksch** [comment]
I merged the latest changes of the `master` - now the functionality works like expected. I tested it for the case that the switch layer is on top level, in a group and even in a sub-group of a group.

File: app/factory/Layer.js [hunk 1]
```diff
@@ -914,87 +914,110 @@ Ext.define('CpsiMapview.factory.Layer', {
...
-                    var wmsFilterString = wmsFilterUtil.getWmsFilterString(newLayer);
+                    if (newLayer.get('isWms')) {
 
-                    // apply new style parameter and reload layer
-                    var newParams = {
-                        STYLES: activeStyle,
-                        FILTER: wmsFilterString,
-                        cacheBuster: Math.random()
-                    };
+                        activeStyle = LegendUtil.getWmsStyleFromSldFile(activeStyle);
```
Comment (geographika) [inline app/factory/Layer.js:967]: Error thrown here as `activeStyle` is undefined. This is when the application starts zoomed in, and the switch layer is off by default, and then zoomed out so the WMS takes over. This happens if all switch layers are switched off and the map is simply zoomed in and out.

**Comment by geographika** [comment]
Thanks @JakobMiksch 

I've tested locally and integrated into a larger project. 

One minor issue in the `mapview` testing - the style radio buttons is not always set/retained when switching layers:

image

Maybe less of an issue if #304 is implemented. 

In the larger project I get an error when zooming in and out. It could be related to the same issue above - the WMS style is not set when the application is started up and the map zoomed in.

**Comment by chrismayer** [comment]
Thanks for your feedback @geographika. I'd pledge to merge this in since this works as expected and create a separate ticket for the remaining minor issue described by you above. Otherwise things feeling a bit messed up because this brings a larger change (larger than expected initially) and maybe has other side-effects with #304.We surely will have a look at the new issue. What do you think?

**Comment by geographika** [comment]
@chrismayer - yes happy to merge. I've never switched branches as much as this week!

**Comment by chrismayer** [comment]
> @chrismayer - yes happy to merge. I've never switched branches as much as this week!

:rofl: : same for me :pray: 

I'll merge then...thanks again for @JakobMiksch and @geographika!

**Pull request merged**